File: mascii2/src/main/java/com/kastkode/mascii2/util/MasciiUtil.java
package com.kastkode.mascii2.util;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import com.kastkode.mascii2.musicelements.Accidental;
import com.kastkode.mascii2.musicelements.KeySignature;
import com.kastkode.mascii2.musicelements.MetaInfo;
import com.kastkode.mascii2.musicelements.Note;
import com.kastkode.mascii2.musicelements.PitchDirection;

public class MasciiUtil {
	final static int PATTERN_INDEX_DEGREE = 3;
	final static int PATTERN_INDEX_ACCIDENTALS = 4;
	final static int PATTERN_INDEX_ABS_PITCH_RANGE = 1;
	final static int PATTERN_INDEX_RANGE_AMPLIFIERS = 2;
	
	public static final String PATTERN_DEGREES = "[a-gA-G]";
	public static final String PATTERN_ACCIDENTALS = "[#@=]";
	public static final Pattern noteSyntax = Pattern.compile("(\\d?)(!*)("+PATTERN_DEGREES+")("+PATTERN_ACCIDENTALS+"*)");

	public static final char REST = '%';
	
	// use a bitmap instead of array to track alterations for each key
	// also needs to handle double sharps, double flats,
	// and even a combo of sharps and flats as it may occur in any measure which 
	// we want to remember for duration of measure
	// we therefore use an int, where:
	//byte 0 (lowest) tracks sharps
	//byte 1 tracks dbl sharps
	//byte 2 tracks flats
	//byte 3 tracks dbl flats

//	public static final Map<String, Integer> circle5thKeyToCode = new HashMap<>(15); 
//	public static final Map<Integer, String> circle5thCodeToKey = new HashMap<>(15); 	

	public static Accidental asAccidental(String accidental) {
		if(accidental == null) return Accidental.NATURAL;
		switch (accidental) {
			case "@@": return Accidental.DBLFLAT;
			case "@": case "=@": return Accidental.FLAT;
			case "#": case "=#": return Accidental.SHARP;
			case "##": return Accidental.DBLSHARP;
			case "=": 
			default: return Accidental.NATURAL;
		}
	}

	public static int magnitude(String shift) {
		if (shift == null) return 0;
		
		return shift.length();
	}

	public static Note.Spelling parsePitch(String pitch) {
		return parsePitch(pitch, KeySignature.C.alterationsMap());
	}

	public static Note.Spelling parsePitch(String pitch, KeySignature.AdHoc key) {
		return parsePitch(pitch, key, null);
	}

	public static Note.Spelling parsePitch(String pitch, Note.Spelling relativeTo) {
		return parsePitch(pitch, KeySignature.C.alterationsMap(),relativeTo);
	}
	
	public static Note.Spelling parsePitch(String pitch, KeySignature.AdHoc key, Note.Spelling relativeTo) {
			//System.out.println("resolving " + pitch + " relative to " + (relativeTo == null? "null" : relativeTo.getPitch()));
		
			Matcher m = MasciiUtil.noteSyntax.matcher(pitch);
			Note.Spelling spelling = null;
			if(m.matches()) {
				spelling = new Note.Spelling(m.group(PATTERN_INDEX_DEGREE).charAt(0));
				String accidentals = m.group(PATTERN_INDEX_ACCIDENTALS);
				String absOctave = m.group(PATTERN_INDEX_ABS_PITCH_RANGE);
				String amplifier = m.group(PATTERN_INDEX_RANGE_AMPLIFIERS);
				
				if(accidentals.length() ==0) {
					spelling.setAccidentals(key.getAccidental(spelling.getDegree()));
				} else {
					spelling.setAccidentals(asAccidental(accidentals));
				}
				if(spelling.isCapitalized()) {
					spelling.setOctaveAdjusts(magnitude(amplifier) * -1);				
				} else {
					spelling.setOctaveAdjusts(magnitude(amplifier));								
				}
	
				//at this point we know degree, up/down shifts, accidentals shift
				//all that remains is what is the basis octave range for this note (before shifts are applied)
				//unless it was specified in absolute terms, we must derive it from the range of the prior note
				//and then go up/down/same from there based on capitalization of this current note's degree
				//same degree and capitalization as previous = unchanged, cap = down, upcase = up
	
				if(absOctave.length() > 0) {  //absolute pitch 					
					spelling.setOctaveRangeBasis(Integer.parseInt(absOctave));
				} else { //relative pitch
					if(relativeTo == null) {
						//relative to nothing, so just assume a default range 
						spelling.setOctaveRangeBasis(
							spelling.isCapitalized()? 3 : 4
						);
					} else {
						spelling.inheritOctaveRangeFrom(relativeTo);
					}
				}
			} else {
				//System.out.println("no match!");
			}
			
			return spelling;			
	}
	
	//the octave boundary is between b and c
	//direction true means upward motion
	//direction false means downward motion
	//same note is considered an octave shift, thus always crosses boundary
	public static boolean crossesOctaveBoundary(char from, char to, PitchDirection direction) {
		from = Character.toLowerCase(from);
		to = Character.toLowerCase(to);

		if(from == to) return direction != PitchDirection.SAME;
		if(direction != PitchDirection.UP) {
			return ! crossesOctaveBoundary(from, to, PitchDirection.UP);
		}
		
		//c to anything cant cross a boundary
		//special case needed because Note.diatonicInterval considers c, c to be unison not octave
		//so math gets skewed
		if(from == 'c') {
			return false;
		}
		
		//calculate upward motion
		return MasciiUtil.diatonicInterval(from, to)
			>=
			MasciiUtil.diatonicInterval(from, 'c');		
	}

	//returns size of diatonic interval.  eg c, e would be 3 (ie its a 3rd)
	//remember that interval math in music counts from 1 not 0, so everything's off by 1
	//a,c should return 3 while c,a should return 6
	public static int diatonicInterval(char A, char B) {
		
		int a = Character.toLowerCase(A) - 'a';
		int b = Character.toLowerCase(B) - 'a';
		
		//java's modulo (%) is incorrect for neg numbers
		//so use this func instead
		return Math.floorMod((b - a), 7) +1 ;		
	}
	
	public static char relativeMajorDegree(char degree) {
		degree = Character.toLowerCase(degree);
		return (char)((((degree - 'a') + 2) % 7) + 'a');
	}
	
	
	public static String[] splitHeaderValues(String content) {
		boolean appendSilentToken = content.matches(".*\\,\\s*$");
		if(appendSilentToken) {
			content += "silentToken";
		}
		String[] headerVals = content.split("\\s*\\,\\s*");
		if(appendSilentToken) {
			headerVals[headerVals.length - 1] = "";
		}
		return headerVals;
	}
	
	public static String times(int howMany, String s) {
		if(howMany <=0) return "";
		if(howMany == 1) return s;
		
		StringBuilder answer = new StringBuilder();
		for(int i=0; i<howMany; i++) {
			answer.append(s);
		}
		return answer.toString();
	}
	
	public static boolean isRest(String s) {
		return s.length() == 1 && s.charAt(0) == REST;
	}
	
	public static List<MetaInfo.Lyric> asLyrics(List<Note> notes, String[] lyrics, String prefix) {
		List<MetaInfo.Lyric> answer = new ArrayList<>(lyrics.length);
		if(lyrics.length == 0) return answer;
		
		int gap = lyrics.length - notes.size(); 
		
		if(gap > 0) {
			//we have too many lyrics, so stuff them into last element
			StringBuilder bldr = new StringBuilder();
			for(int i=lyrics.length - gap - 1; i < lyrics.length; i++) {
				String lyr = lyrics[i];
				if(! isRest(lyr)) {					
					bldr.append(lyr);
					if(i + 1<lyrics.length) {
						bldr.append("-");
					}
				}
			}
			lyrics[lyrics.length - gap - 1] = bldr.toString();
		}
		
		int lind = 0;			
		for(Iterator<Note> it = notes.iterator(); it.hasNext() && lind < lyrics.length; lind++) {
			Note n = it.next();
			String lyr = lyrics[lind];
			if(isRest(lyr)) {
				answer.add(null);
				continue;
			}
			MetaInfo.Lyric lyric = new MetaInfo.Lyric();
			lyric.setStartingAt(n.getStart());
			lyric.setRawValue(prefix + lyr);
			answer.add(lyric);
		}
		return answer;
	}
	
	//assumes ArrayLists
	public static void mergeBIntoA(List<MetaInfo.Lyric> a, List<MetaInfo.Lyric> b ) {		
		for(int ai=0, bi=0; ai < a.size() && bi < b.size(); ai++, bi++) {
			MetaInfo.Lyric la = a.get(ai);
			MetaInfo.Lyric lb = b.get(bi);			
			
			if(la == null) {
				a.set(ai, lb);
			} else if(lb == null) {
				
			} else {
				la.setRawValue(la.getRawValue() + lb.getRawValue());
			}
		}
		
		int sizeDiff = b.size() - a.size();
		if(sizeDiff > 0) {
			a.addAll(b.subList(b.size() - sizeDiff - 1, b.size()));
		}
	}
}

File: mascii2-typescript/src/util/MasciiUtil.ts
import { Accidental } from '../musicelements/Accidental';
import { KeySignatureAdHoc } from '../musicelements/KeySignature';
import { Note, NoteSpelling } from '../musicelements/Note';
import { MetaInfo } from '../musicelements/MetaInfo';
import { PitchDirection } from '../musicelements/PitchDirection';
import { Lyric } from '../musicelements/MetaInfo';
import { PitchContext } from '../antlr-generated/MasciiParser';

export const REST = '%';

export function asAccidental(accidental: string | undefined | null): Accidental {
    if (!accidental) return Accidental.NATURAL;
    switch (accidental) {
        case '@@': return Accidental.DBLFLAT;
        case '@': case '=@': return Accidental.FLAT;
        case '#': case '=#': return Accidental.SHARP;
        case '##': return Accidental.DBLSHARP;
        default: return Accidental.NATURAL;
    }
}

export function magnitude(shift: string | undefined | null): number {
    if (!shift) return 0;
    return shift.length;
}

export function parsePitchContext(ctx: PitchContext, key: KeySignatureAdHoc, rel?: NoteSpelling | null): NoteSpelling {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const repeatEl = ctx.REPEAT_ELEMENT() as any;
    if (repeatEl != null) {
        return new NoteSpelling(repeatEl.getText() as string);
    }
    const absOctave: string = (ctx.abs_pitch_range() as unknown as { getText(): string } | null)?.getText() ?? '';
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const amplifier: string = (ctx.AMPLIFIERS() as any)?.getText() ?? '';
    const degree = ctx.REL_PITCH().getText();
    const accidentals: string = (ctx.accidental() as unknown as { getText(): string } | null)?.getText() ?? '';

    const spelling = new NoteSpelling(degree);

    if (accidentals.length === 0) {
        spelling.accidentals = key.getAccidental(degree);
    } else {
        spelling.accidentals = asAccidental(accidentals);
    }

    if (spelling.isCapitalized()) {
        spelling.octaveAdjusts = magnitude(amplifier) * -1;
    } else {
        spelling.octaveAdjusts = magnitude(amplifier);
    }

    if (absOctave.length > 0) {
        spelling.setOctaveRangeBasis(parseInt(absOctave, 10));
    } else {
        if (rel == null) {
            spelling.setOctaveRangeBasis(spelling.isCapitalized() ? 3 : 4);
        } else {
            spelling.inheritOctaveRangeFrom(rel);
        }
    }

    return spelling;
}

export function crossesOctaveBoundary(from: string, to: string, direction: PitchDirection): boolean {
    const f = from.toLowerCase();
    const t = to.toLowerCase();
    if (f === t) return direction !== PitchDirection.SAME;
    if (direction !== PitchDirection.UP) {
        return !crossesOctaveBoundary(f, t, PitchDirection.UP);
    }
    if (f === 'c') return false;
    return diatonicInterval(f, t) >= diatonicInterval(f, 'c');
}

export function diatonicInterval(a: string, b: string): number {
    const ai = a.toLowerCase().charCodeAt(0) - 'a'.charCodeAt(0);
    const bi = b.toLowerCase().charCodeAt(0) - 'a'.charCodeAt(0);
    return ((((bi - ai) % 7) + 7) % 7) + 1;
}

export function relativeMajorDegree(degree: string): string {
    const d = degree.toLowerCase();
    return String.fromCharCode((((d.charCodeAt(0) - 'a'.charCodeAt(0)) + 2) % 7) + 'a'.charCodeAt(0));
}

export function splitHeaderValues(content: string): string[] {
    const appendSilent = /,\s*$/.test(content);
    let s = content;
    if (appendSilent) s += 'silentToken';
    const parts = s.split(/\s*,\s*/);
    if (appendSilent) parts[parts.length - 1] = '';
    return parts;
}

export function times(howMany: number, s: string): string {
    if (howMany <= 0) return '';
    return s.repeat(howMany);
}

export function isRest(s: string): boolean {
    return s.length === 1 && s === REST;
}

export function asLyrics(notes: Note[], lyrics: string[], prefix: string): (Lyric | null)[] {
    const answer: (Lyric | null)[] = [];
    if (lyrics.length === 0) return answer;

    const gap = lyrics.length - notes.length;
    if (gap > 0) {
        const sb: string[] = [];
        for (let i = lyrics.length - gap - 1; i < lyrics.length; i++) {
            const lyr = lyrics[i]!;
            if (!isRest(lyr)) {
                sb.push(lyr);
                if (i + 1 < lyrics.length) sb.push('-');
            }
        }
        lyrics[lyrics.length - gap - 1] = sb.join('');
    }

    let lind = 0;
    for (const n of notes) {
        if (lind >= lyrics.length) break;
        const lyr = lyrics[lind]!;
        if (isRest(lyr)) {
            answer.push(null);
            lind++;
            continue;
        }
        const lyric = new Lyric();
        lyric.setStartingAt(n.getStart());
        lyric.setRawValue(prefix + lyr);
        answer.push(lyric);
        lind++;
    }
    return answer;
}

export function mergeBIntoA(a: (Lyric | null)[], b: (Lyric | null)[]): void {
    for (let ai = 0, bi = 0; ai < a.length && bi < b.length; ai++, bi++) {
        const la = a[ai];
        const lb = b[bi];
        if (la == null) {
            a[ai] = lb ?? null;
        } else if (lb != null) {
            la.setRawValue((la.getRawValue() ?? '') + (lb.getRawValue() ?? ''));
        }
    }
    const sizeDiff = b.length - a.length;
    if (sizeDiff > 0) {
        const tail = b.slice(b.length - sizeDiff - 1);
        for (const item of tail) a.push(item);
    }
}
